Deduplicate rules prior to run (Closes: #2298)

hlieberman-rd opened the pull request:
Because the rule objects contain context from the ruleset they came
from, we can't directly compare them. Instead, we have to do a little
dance and indirectly deduplicate them by their id.

Signed-off-by: Harlan Lieberman-Berg <harlan@rebelliondefense.com>

hlieberman-rd committed:
Deduplicate rules prior to run (Closes: #2298)

Because the rule objects contain context from the ruleset they came
from, we can't directly compare them. Instead, we have to do a little
dance and indirectly deduplicate them by their id.

Signed-off-by: Harlan Lieberman-Berg <harlan@rebelliondefense.com>

hlieberman-rd commented:
This saves about 10% on the intersection of r2c-ci and r2c-best-practices (234 deduped vs 259).

aryx commented:
CI is failing. Needs investigation.
You can run the tests locally with: cd /path/to/semgrep_repo/semgrep; make test

aryx reviewed (changes requested):
Failure in CI.
There's also probably something shorter than this in Python. @mschwager ?

hlieberman-rd commented:
The CI failures appear unrelated; origin/develop is failing on my machine
as well.--
Harlan Lieberman-Berg
Head of Security
Rebellion Defense

aryx commented:
I don't think this one is failing in develop: FAILED tests/e2e/test_check.py::test_multiple_configs_file - AssertionError: ...

But yes, thx for reporting I'll try to fix develop.

aryx commented:
@cla-bot check

aryx commented:
@cla-bot check

aryx commented:
Yeah, the verification/cla-signed approved it!

mschwager commented on semgrep/semgrep/config_resolver.py:
```diff
@@ -133,7 +133,16 @@ def get_rules(self, no_rewrite_rule_ids: bool) -> List[Rule]:
+        rules = []
+        seen = set()
+        for rule in [rule for rules in configs.values() for rule in rules]:
+            if rule.id in seen:
```
Rule IDs are not a one-to-one mapping with a rule ([see e.g. here](https://github.com/returntocorp/semgrep-rules/blob/develop/go/lang/correctness/permissions/file_permission.yaml)), so we can't use `rule.id` to filter seen rules. If we define a reasonable `__eq__` comparator on the `rule.Rule` class, then we could simply do something like the following here:

```
return set(rule for rules in configs.values() for rule in rules)
```

And the `set` will handle everything for us :+1:

hlieberman-rd replied:
Hi @mschwager!

Good to know; I will have to extend Rule with `__eq__` and `__hash__`.

Unfortunately, we can't use sets here for the rules themselves; the order being non-deterministic causes serious problems with the snapshot tests.  I talked to @brendongo about this on Slack, and the best option that we came to was to continue maintaining them as a list for the moment pending a more significant overhaul of this section of the code.

I experimented with passing it through `dict.fromkeys`, however, with Python 3.6, we'd be relying on a Cython implementation detail; dict was not guaranteed to preserve insertion order until Python 3.7.

The other option is to pass it through set as an intermediary and then sort the output, however, that would require adding `__lt__` and regenerating all the tests.

Happy to do either; let me know which is preferable.

mschwager replied:
Hey there,

Ahh, interesting, good catch. We could use `OrderedDict.fromkeys` - something like:

```python
# Remove duplicates, preserve order
return list(OrderedDict.fromkeys(rule for rules in configs.values() for rule in rules))
```

Would that minimize snapshot thrashing?

hlieberman-rd committed:
Revise rule deduplication.

This patch also provides an implementation of __eq__ and __hash__ for
the Rule type.

Signed-off-by: Harlan Lieberman-Berg <harlan@rebelliondefense.com>

brendongo commented:
Fixed issue on develop and rerunning tests

nbrahms commented:
@aryx is your review still pertinent, or should I dismiss it? 🙏

aryx commented:
There's still this CI failure: FAILED tests/e2e/test_check.py::test_multiple_configs_file - AssertionError: ...
 that does not happen on develop.
We can't merge until all the tests pass.

mschwager commented on semgrep/semgrep/rule.py:
```diff
@@ -68,6 +68,15 @@ def __init__(self, raw: YamlTree[YamlMap]) -> None:
+        return self._raw == other._raw
+
+    def __hash__(self) -> int:
+        return hash(self.id)
```
I think this may be problematic given the above definition of `__eq__`. Since `id` is not guaranteed to be unique this means we could have rules that produce the same hashable value, but are not `==`. I.e. they overwrite each other in a `dict` or `set`, but are not equivalent. If we want something really quick and easy, although not particularly fast, we could do something like `hash(json.dumps(self._raw))` or something like that.

hlieberman-rd replied:
@mschwager That's OK -- hash equivalence isn't treated as the same as equality.

For example:

```python3
class Foo:
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def __eq__(self, other):
        if not isinstance(other, type(self)):
            return False
        return (self.a == other.a and self.b == other.b)

    def __hash__(self):
        return hash(self.a)

x = Foo("abc", "def")
y = Foo("abc", "xyz")
z = Foo(888, "def")

print(hash(x))
print(hash(y))
print(hash(z))
print(x == y)
print(x == z)
print(y == z)
```
returns
```
-4144991442677912395
-4144991442677912395
888
False
False
False
```

Equivalence never checks `__hash__`.  Membership does (a la `x in y`), but only as a performance improvement.  Hash collisions are, afaik, only a performance hit, never a logical mistake.  (The performance hit of doing the deep comparison between the raw dicts on id collision is less than the hashing and serializing the json on every comparison).

I believe the test failure you're seeing @aryx is actually the corrected behavior: the snapshot version appears to be producing two hits from identical rules in two files.

```
>>> with open("tests/e2e/snapshots/test_check/test_multiple_configs_file/results.json") as x:
...     y = x.read()
...
>>> import json
>>> foo = json.loads(y)
>>> foo["results"][2] == foo["results"][0]
True
```

I will push a new patch correcting the snapshot.

hlieberman-rd replied:
Sorry @mschwager, I misread your concern.  dicts and sets also check equality, not just hashing.  Extending the above Foo class example:

```
bar = set()
bar.add(x)
bar.add(y)
bar.add(z)
print(len(bar))
bar.add(x)
bar.add(y)
print(len(bar))
```
returns
```
3
3
```

mschwager replied:
Ahh, TIL that hashable types rely on `__eq__` too! [`__hash__`](https://docs.python.org/3/reference/datamodel.html#object.__hash__) was a lengthy, but helpful read. Thanks for the clarifications! This LGTM :+1:

hlieberman-rd committed:
Commit now-passing test results.

Signed-off-by: Harlan Lieberman-Berg <harlan@rebelliondefense.com>

brendongo commented:
Thanks @hlieberman-rd this should go in with next week's release!
